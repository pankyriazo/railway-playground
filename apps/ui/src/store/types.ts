import { Deployment, Edges, Projects } from "@monorepo/types";

export type State = {
  token: string;
  projects: Projects;
  deployments: Record<string, Edges<Deployment>>;
};

interface ActionBase {
  type: string;
  payload?: unknown;
}

interface ResetAction extends ActionBase {
  type: "RESET";
}
interface SetTokenAction extends ActionBase {
  type: "SET_TOKEN";
  payload: string;
}
interface SetProjectsAction extends ActionBase {
  type: "SET_PROJECTS";
  payload: Projects;
}
interface AddDeploymentAction extends ActionBase {
  type: "ADD_DEPLOYMENT";
  payload: Record<string, Edges<Deployment>>;
}
interface RemoveDeploymentAction extends ActionBase {
  type: "REMOVE_DEPLOYMENT";
  payload: string;
}

export type Action =
  | ResetAction
  | SetTokenAction
  | SetProjectsAction
  | AddDeploymentAction
  | RemoveDeploymentAction;

export type Listener = (state: State) => void;

export type Reducer = (state: State, action: Action) => State;
