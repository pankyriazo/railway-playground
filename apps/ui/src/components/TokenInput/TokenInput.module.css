.container {
  display: flex;
  flex-direction: column;
  margin: var(--spacing-lg) auto;

  max-width: 400px;
}

.label {
  color: var(--color-text--700);
  font-size: var(--font-size-xs);
}

.input {
  margin-top: var(--spacing-xs);
}

.btn {
  composes: btn from global;

  margin-top: var(--spacing-sm);
}
