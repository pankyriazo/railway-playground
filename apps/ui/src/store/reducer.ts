import { Action, State } from "./types";
import { initialState } from "./store";

const reducer = (state: State, action: Action): State => {
  switch (action.type) {
    case "RESET":
      return initialState;
    case "SET_TOKEN":
      return { ...state, token: action.payload };
    case "SET_PROJECTS":
      return { ...state, projects: action.payload };
    case "ADD_DEPLOYMENT":
      return {
        ...state,
        deployments: { ...state.deployments, ...action.payload },
      };
    case "REMOVE_DEPLOYMENT":
      const { [action.payload]: _, ...deployments } = state.deployments;
      return { ...state, deployments };
    default:
      return state;
  }
};

export default reducer;
