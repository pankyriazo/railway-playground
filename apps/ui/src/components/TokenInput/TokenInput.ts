import store from "@/store";
import { Component } from "@/types";

import styles from "./TokenInput.module.css";

class TokenInput extends Component<HTMLFormElement> {
  private _token: string = "";

  constructor() {
    super();

    this._element = document.createElement("form");
    this._element.id = "form";
    this._element.addEventListener("submit", this._handleSubmit.bind(this));

    this._subscribe();
    this._checkLocalStorage();

    this.render();
  }

  private _checkLocalStorage() {
    const token = localStorage.getItem("token");

    if (token) {
      store.dispatch({
        type: "SET_TOKEN",
        payload: token,
      });
    }
  }

  private _subscribe() {
    return store.subscribe((state) => {
      if (state.token !== this._token) {
        this._token = state.token;

        if (state.token === "") {
          localStorage.removeItem("token");
        } else {
          localStorage.setItem("token", this._token);
        }

        this.render();
      }
    });
  }

  private _handleSubmit(event: Event) {
    event.preventDefault();

    if (!this._element) {
      return;
    }

    const token = (this._element.querySelector("#token") as HTMLInputElement)
      .value;

    store.dispatch({
      type: "SET_TOKEN",
      payload: token,
    });
  }

  override render() {
    if (!this._element) {
      return;
    }

    this._element.innerHTML = "";
    this._element.dataset.testid = "token-input__form";

    const containerElement = document.createElement("div");
    containerElement.className = styles.container!;

    const labelElement = document.createElement("label");
    labelElement.dataset.testid = "token-input__label";
    labelElement.htmlFor = "token";
    labelElement.textContent = "Access token";
    labelElement.className = styles.label!;
    containerElement.appendChild(labelElement);

    const inputElement = document.createElement("input");
    inputElement.dataset.testid = "token-input__input";
    inputElement.type = "password";
    inputElement.id = "token";
    inputElement.className = styles.input!;
    inputElement.value = this._token;
    inputElement.placeholder = "Enter your access token";
    inputElement.required = true;
    inputElement.oninvalid = () => {
      inputElement.setCustomValidity("Please enter your access token");
    };
    inputElement.oninput = () => {
      inputElement.setCustomValidity("");
    };
    if (this._token === "") {
      inputElement.autofocus = true;
    }
    containerElement.appendChild(inputElement);

    const buttonElement = document.createElement("button");
    buttonElement.dataset.testid = "token-input__submit";
    buttonElement.type = "submit";
    buttonElement.textContent = this._token ? "Update token" : "Use token";
    buttonElement.className = styles.btn!;
    containerElement.appendChild(buttonElement);

    this._element.appendChild(containerElement);
  }
}

export default TokenInput;
