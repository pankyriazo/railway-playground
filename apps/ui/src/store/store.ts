import reducer from "./reducer";
import { Action, Listener, Reducer, State } from "./types";

export const initialState: State = {
  token: "",
  projects: {
    data: null,
    errors: [],
  },
  deployments: {},
};

class Store {
  private _state: State;
  private _reducer: Reducer;
  private _listeners: Listener[] = [];

  constructor() {
    this._state = initialState;
    this._reducer = reducer;
  }

  get state(): State {
    return this._state;
  }

  private _notify(): void {
    this._listeners.forEach((listener) => listener(this._state));
  }

  dispatch(action: Action): void {
    this._state = this._reducer(this._state, action);
    this._notify();
  }

  subscribe(listener: Listener): () => void {
    this._listeners.push(listener);

    listener(this._state);

    return () => {
      this._listeners = this._listeners.filter((l) => l !== listener);
    };
  }
}

const store = new Store();

export default store;
