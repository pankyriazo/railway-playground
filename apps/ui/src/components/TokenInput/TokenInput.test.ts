import { render } from "@/test-helpers";
import store from "@/store";
import TokenInput from "./TokenInput";

describe("TokenInput", () => {
  beforeEach(() => {
    render(TokenInput);
  });

  afterEach(() => {
    store.dispatch({ type: "RESET" });
  });

  it("should render", () => {
    expect(document.body).toMatchSnapshot();
  });

  it("should render the correct label", () => {
    const label = document.querySelector("label");

    expect(label).not.toBeNull();
    expect(label!.textContent).toBe("Access token");
  });

  it("should render the input as password type", () => {
    const input = document.querySelector("input");

    expect(input!.getAttribute("type")).toBe("password");
  });

  describe("submit button", () => {
    it("should render", () => {
      const button = document.querySelector("button");

      expect(button).not.toBeNull();
    });

    it("should render the correct text when the token is empty", () => {
      const button = document.querySelector("button");

      expect(button!.textContent).toBe("Use token");
    });

    it("should render the correct text when the token is not empty", () => {
      store.dispatch({
        type: "SET_TOKEN",
        payload: "test-token",
      });

      render(TokenInput);

      const button = document.querySelector("button");

      expect(button!.textContent).toBe("Update token");
    });
  });

  describe("on submit button click", () => {
    let storeSpy: jest.SpyInstance;

    beforeEach(() => {
      storeSpy = jest.spyOn(store, "dispatch");
    });

    afterEach(() => {
      storeSpy.mockRestore();
    });

    describe("when the token is empty", () => {
      it("should not dispatch the SET_TOKEN action", () => {
        const submitButton = document.querySelector("button");
        submitButton!.click();

        expect(storeSpy).not.toHaveBeenCalled();
      });
    });

    describe("when the token is not empty", () => {
      it("should dispatch the SET_TOKEN action", () => {
        const input = document.querySelector("input");
        input!.value = "token";
        input!.dispatchEvent(new Event("input"));

        const submitButton = document.querySelector("button");
        submitButton!.click();

        expect(storeSpy).toHaveBeenCalledTimes(1);
        expect(storeSpy).toHaveBeenCalledWith({
          type: "SET_TOKEN",
          payload: "token",
        });
      });
    });
  });
});
