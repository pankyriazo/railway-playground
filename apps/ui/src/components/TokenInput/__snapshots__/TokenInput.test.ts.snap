// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenInput should render 1`] = `
<body>
  <form
    data-testid="token-input__form"
    id="form"
  >
    <div
      class="container"
    >
      <label
        class="label"
        data-testid="token-input__label"
        for="token"
      >
        Access token
      </label>
      <input
        autofocus=""
        class="input"
        data-testid="token-input__input"
        id="token"
        placeholder="Enter your access token"
        required=""
        type="password"
      />
      <button
        class="btn"
        data-testid="token-input__submit"
        type="submit"
      >
        Use token
      </button>
    </div>
  </form>
</body>
`;
